hibernate-core/src/main/java/org/hibernate/sql/results/spi/ListResultsConsumer.java

```diff
@@ -89,19 +90,60 @@ public enum UniqueSemantic {
 	}
 
 	private final UniqueSemantic uniqueSemantic;
-	private final ResultHandler<R> resultHandler;
 
 	public ListResultsConsumer(UniqueSemantic uniqueSemantic) {
 		this.uniqueSemantic = uniqueSemantic;
+	}
+
+	private static class Results<R> {
+		private final List<R> results = new ArrayList<>();
+		private final JavaType resultJavaType;
+
+		public Results(JavaType resultJavaType) {
+			this.resultJavaType = resultJavaType;
+		}
+
+		private boolean contains(R result) {
+			for ( int i = 0; i < results.size(); i++ ) {
+				if ( resultJavaType.areEqual( results.get( i ), result ) ) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		public boolean addUnique(R result) {
+			if ( contains( result ) ) {
+				return false;
+			}
```

Comment:
Small nitpick, I'd prefer to inline the `contains` method.

```suggestion
			for ( int i = 0; i < results.size(); i++ ) {
				if ( resultJavaType.areEqual( results.get( i ), result ) ) {
					return false;
				}
			}
```